### ISSUE TITLE
Automatic registration for organizers and Delegates

### ISSUE THREAD

**Issue by viroulep** [thewca/worldcubeassociation.org#1692]
Right now it could happen that a Delegate or Organizer is put on waiting list because of a registration limit (eg: if registrations fill very very quickly, and the Delegate/Organizer does not register quickly enough).

Since they literally create the competition and choose registration's opening/closing, I think it would make sense to automatically create a registration for them upon competition's confirmation.
I quickly talked about this with @Laura-O who thinks it makes sense too.

**Comment by Mollerz** [comment]
I was thinking about something like this as well, for delegates or organisers that are not competing then it is not a problem to remove them from the registration list. I think something like that happens far less often as well, and it's very easy to delete a registration. Good change in my opinion.

**Comment by cubewhiz** [comment]
I think this is even more critical now that a delegate recently could not compete because he did not register for his own competition.

**Comment by jfly** [comment]
When you automatically register a delegate for a competition, what events should you register him/her for?

**Comment by cubewhiz** [comment]
How about that person's profile preferences?

If they have no preferences, can you register for no events? Is that possible? They could amend their registration later, but at least they would be on the list.

**Comment by jfly** [comment]
Currently, we do not allow people to register for 0 events. We could change that, but I don't think we actually do want people creating registrations for 0 events...

**Comment by cubewhiz** [comment]
How about if they can't be bothered to fill out their preference in their profile, then they get registered for nothing? :P 

I agree, don't change to allow registration for 0 events.

**Comment by Baiqiang** [comment]
I don't really like this idea. It lacks fair for others. Every competitor including delegates and organizers must use the same registration process. In China, Fangyuan, Xiaobo and Chirs have missed the registration several times and didn't compete finally.
I agree with delegates/organizers choosing the registration's opening time. But I think we must announce it enough time like 3 days so that community know it and everyone gets an equal chance to register. Especially when there's a limit for competitors, "first come first serve" basis must be applied on everyone.

**Comment by cubewhiz** [comment]
Isn't it more unfair that the person who is organizing or delegating the
competition doesn't have the opportunity to compete? I don't think this
person is violating the "first come first serve" basis because they are
literally the first one--they are creating the page in the first place!


Bob Burton

**Comment by Baiqiang** [comment]
The registration process is the same for everyone, regardless their status. This is inherently fair. The organizer knows when registration begins, they can make a point to be sure to register on time.

**Comment by jfly** [comment]
The Board just informed the WST that they are ok with the WST implementing this feature.

**Comment by Baiqiang** [comment]
I still can't understand why the organizers/delegates have privilege on registration. We just released the new regulations and added some restrictions on the registration. Why must we grant delegates/organizers the privilege?

**Comment by moralsh** [comment]
I might be interested in implementing this (if no one tells me I'm crazy), What I understad that has to be done:

- Create one or two checkboxes (default checked) on the new comp form to automatically register delegates and/or organizers
- Create one or two boolean columns on Competitions table to reflect that
- If checked, add registration in the default profile events, if there are no default events, don't register and maybe issue a warning or notify the delegate creating the competition.

With the checkboxes we're letting the delegate creating the competition decide on giving the privilege to themselves and fellow organizers although I also can't see what's wrong with making sure the people running the comp compete.

Possible problems: 

I don't know  if an extra organizer/delegate can be added after creating the comp, as I am no delegate and haven't clicked everywhere on my dev server yet, but if they can be added we have to also take care of the registrations while editing

**Comment by jfly** [comment]
Thanks, Raul! I don't like the complication of giving an option for this, let's just automatically register them. Yes, delegates and organizers can be added after a competition is announced, and we should definitely handle this case! I think it could work to put all this logic inside of the `CompetitionDelegate` and `CompetitionOrganizer` models in an `after_create` callback.

**Comment by cubewhiz** [comment]
I think a fair compromise is to have the delegate and organizers automatically registered, but not to automatically confirm that registration...or maybe that just makes this needlessly more complicated.

As for handling the case of delegates and organizers to be added after the competition is announced, I would like to clarify that those people should NOT be automatically registered as this could just lead to an abuse of this process.

**Comment by jfly** [comment]
> have the delegate and organizers automatically registered, but not to automatically confirm that registration...or maybe that just makes this needlessly more complicated.

I'd rather not do that. A lot of delegates have asked us to implement automatic confirmation of registrations, and whenever we do get around to working on that, having delegates/organizers automatically end up in this unconfirmed state is almost certainly going to cause issues.

> As for handling the case of delegates and organizers to be added after the competition is announced, I would like to clarify that those people should NOT be automatically registered as this could just lead to an abuse of this process.

Ah, you're absolutely right.

**Comment by moralsh** [comment]
> As for handling the case of delegates and organizers to be added after the competition is announced, I would like to clarify that those people should NOT be automatically registered as this could just lead to an abuse of this process.

Just curious to learn How the process can be abused. To register in an already full competition? I'd rather put checks on this or other things than not registering, but maybe I'm not getting what you are meaning.

@jfly If we don't automatically register organizers and delegates added after the competition is announced as @cubewhiz suggests, Do you still see the callback after the object creation as the best place to do this adding checks there? (just trying to get used to how you guys do things here)

**Comment by cubewhiz** [comment]
Indeed, the abuse would come by means of adding a delegate/organizer when the competition is already full. For example, my friend did not register in time, so now I add them as an organizer and now they are registered for the competition.

If the competition is not full, though, then I see no reason why you couldn't add delegates/organizers automatically.

**Comment by jfly** [comment]
@moralsh, how about automatically registering delegates and organizers so long as one of the two is true:

1. The competitor limit is disabled
2. The competitor limit is enabled, **and** registration is not full yet.

**Comment by Mollerz** [comment]
Bringing this up again, our latest competition registration filled up in only an hour or so and one organiser registered with only 2 free spaces left as he was busy during the time it opened. Another organiser had to pull over whilst driving to register to get in...

**Comment by jfly** [comment]
Actually, on further thought, I don't think this is necessarily going to be work very well as we've planned it out here. Here's what I imagine is the normal flow:

1. Competition is created, it has no events.
2. Organizers and delegates are added.
3. Later on, competition events are added.

If we try to register organizers and delegates when they're added (in step 2), then there won't be any events to register them for! I don't know what the right way to handle this is.

**Comment by jfly** [comment]
(cc-ing @ronaldmansveld as this could be a good thing for him to think about if he has any time to meditate on our whole registration process)

**Comment by AlbertoPdRF** [comment]
Maybe register them when the competition gets confirmed or announced?

**Comment by jfly** [comment]
Registering when the competition is confirmed sounds pretty reasonable to me!

**Comment by ronaldmansveld** [comment]
1) we can register them when the competition is confirmed, as then all
events etc get locked anyway, preventing more issues when these get changed
2) we can register them when registration opens. I feel this seems more
logical for the outside, as otherwise they may get registered before
registration is open, leading to questions from regular competitors.

As for events: can we circumvent the event check for this automatic
registration in case they have no preferred events set? Otherwise we can
default to either just 3x3, or all events they have competed in in the last
year?


Ronald

**Comment by AlbertoPdRF** [comment]
The @thewca/board hasn't discussed this again as far as I remember, but with https://github.com/thewca/wca-documents/pull/40 I'm afraid this won't technically be possible, so maybe this issue should be closed?

**Comment by cubewhiz** [comment]
We discussed and agreed on this over a year ago. I'd rather see this implemented than closed, so if it means editing the Competition Requirements, I'd prefer that route. What in that document do you think prohibits this?

**Comment by dunkOnIT** [comment]
Leaving this open as implementing any kind of auto-accept feature could cause issues with organisers/delegates not being accepted if registrations can only be approved once registration opens. 

* System is implemented to automatically approve registrations, which approves registrations in order of when payment is received
* If organisers/delegates forget to pay and the comp fills up with auto-accepted paid registrations in a minute, they won't have a spot
* This could be avoided by organisers/delegates paying in advance (thus their payments would be the earliest)
* We'd need logic that looks at the list of payments to accept, though, rather than acts when it receives notification that a payment has occurred (otherwise payments happening before the auto-accept agent is listening would then be ignored)

**Comment by dunkOnIT** [comment]
Closing as our solution for this is that organizers and delegates must pre-register and accept their registrations before reg opens - as of today, that is supported in WCRP :)
